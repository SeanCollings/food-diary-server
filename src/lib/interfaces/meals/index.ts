export enum EMealType {
  BREAKFAST = 'breakfast',
  SNACK_1 = 'snack_1',
  LUNCH = 'lunch',
  SNACK_2 = 'snack_2',
  DINNER = 'dinner',
}
export type TMealType = `${EMealType}`;

export type TSelectedEmoji = {
  name: string;
  nativeSkin: string;
} | null;

export interface IMealContent {
  id: string;
  emoji?: TSelectedEmoji;
  serving?: string;
  measurement?: string;
  food: string;
  description?: string;
}

export interface IMealTypeContent {
  contents: IMealContent[];
}

export type TMealCard = {
  [key in TMealType]?: IMealTypeContent;
};
