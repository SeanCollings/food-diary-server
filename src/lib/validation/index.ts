import { trim } from '@/utils/string-utils';
import {
  EMAIL_MIN_LENGTH,
  EMAIL_REGEX,
  INPUT_MAX_LENGTH,
  PASSWORD_MIN_LENGTH,
} from '@/lib/constants/validation/validation.constants';

export const validateEmail = (email: string) => {
  const trimEmail = trim(email);

  if (!trimEmail?.length) {
    return false;
  } else if (!EMAIL_REGEX.test(trimEmail)) {
    return false;
  }

  return (
    trimEmail.length >= EMAIL_MIN_LENGTH && trimEmail.length <= INPUT_MAX_LENGTH
  );
};

export const validatePassword = (password: string) => {
  const trimPassword = trim(password);

  if (!trimPassword?.length) {
    return false;
  }

  return (
    trimPassword.length >= PASSWORD_MIN_LENGTH &&
    trimPassword.length <= INPUT_MAX_LENGTH
  );
};

export const validateLogin = ({
  email,
  password,
}: {
  email: string;
  password: string;
}) => validateEmail(email) && validatePassword(password);
