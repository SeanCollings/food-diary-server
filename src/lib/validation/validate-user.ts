import { promisify } from 'util';
import { scrypt as _scrypt } from 'crypto';
import { User } from '@prisma/client';
import { SCRYPT_KEYLEN } from '@/auth/auth.constants';

const scrypt = promisify(_scrypt);

export const isValidPassword = async (user: User | null, password: string) => {
  if (!user) {
    return null;
  }

  const [salt, storedHash] = user.password.split('.');
  const hash = (await scrypt(password, salt, SCRYPT_KEYLEN)) as Buffer;

  if (storedHash === hash.toString('hex')) {
    return user;
  }

  return null;
};
