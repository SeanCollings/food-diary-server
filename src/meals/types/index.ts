import { IsDateString, IsNotEmpty } from 'class-validator';
import { Request } from 'express';

// https://stackoverflow.com/questions/50160662/convert-nested-array-type

export class CreateMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class UpdateMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class DeleteMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export enum MealTypesEnum {
  BREAKFAST = 'breakfast',
  SNACK_1 = 'snack_1',
  LUNCH = 'lunch',
  SNACK_2 = 'snack_2',
  DINNER = 'dinner',
}
export type MealTypes = `${MealTypesEnum}`;

export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export interface MealContent {
  id: string;
  food: string;
  serving?: string;
  quantity?: string;
  description?: string;
  emoji?: { name: string; nativeSkin: string } | null;
}

export type SetFieldOld<V, O extends object> = {
  [K in keyof O]: O[K] extends object ? SetFieldOld<V, O[K]> : V;
};

type Primitive = string | number | boolean | null | undefined;

type SetField<V, O> = O extends Primitive
  ? V
  : O extends (infer U)[]
  ? SetField<V, U>
  : { [K in keyof O]: SetField<V, O[K]> };

type RequireOne<T> = T & { [P in keyof T]: Required<Pick<T, P>> }[keyof T];
