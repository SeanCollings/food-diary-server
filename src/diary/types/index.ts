import { MAX_CALENDAR_ENTRIES_RANGE } from '@/lib/constants/validation/validation.constants';
import { Type } from 'class-transformer';
import { IsDateString, IsNotEmpty, IsNumber, Max } from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export class GetDiaryEntriesQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class GetCalendarEntriesQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;

  @IsNotEmpty()
  @IsNumber()
  @Max(MAX_CALENDAR_ENTRIES_RANGE)
  @Type(() => Number)
  months: string;
}

export type MealColumnKeys =
  | 'mealBreakfast'
  | 'mealSnack1'
  | 'mealLunch'
  | 'mealSnack2'
  | 'mealDinner';

export type HasMealColumnKeys =
  | 'hasMealBreakfast'
  | 'hasMealSnack1'
  | 'hasMealLunch'
  | 'hasMealSnack2'
  | 'hasMealDinner';
