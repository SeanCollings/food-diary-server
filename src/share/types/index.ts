import { MAX_SUMMARY_MONTH_RANGE } from '@/lib/constants/validation/validation.constants';
import { IsDateAfter } from '@/lib/validation/validators/is-date-after';
import { IsWithinDateRange } from '@/lib/validation/validators/is-within-date-range';
import {
  IsDateString,
  IsNotEmpty,
  IsString,
  IsUUID,
  MaxLength,
  MinLength,
} from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export class GetSharedSummaryQuery {
  @IsNotEmpty()
  @MinLength(36)
  @MaxLength(36)
  @IsString()
  @IsUUID()
  link: string;

  @IsNotEmpty()
  @IsDateString()
  dateFrom: string;

  @IsNotEmpty()
  @IsDateString()
  @IsDateAfter('dateFrom', { message: 'dateFrom must come before dateTo' })
  @IsWithinDateRange('dateFrom', {
    message: `dateFrom and dateTo max-range is ${MAX_SUMMARY_MONTH_RANGE} months`,
  })
  dateTo: string;
}
