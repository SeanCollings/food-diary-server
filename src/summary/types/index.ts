import { MAX_SUMMARY_MONTH_RANGE } from '@/lib/constants/validation/validation.constants';
import { IsDateAfter } from '@/lib/validation/validators/is-date-after';
import { IsWithinDateRange } from '@/lib/validation/validators/is-within-date-range';
import { IsDateString, IsNotEmpty } from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export class GetUserSummaryQuery {
  @IsDateString()
  @IsNotEmpty()
  dateFrom: string;

  @IsDateString()
  @IsNotEmpty()
  @IsDateAfter('dateFrom', { message: 'dateFrom must come before dateTo' })
  @IsWithinDateRange('dateFrom', {
    message: `dateFrom and dateTo max-range is ${MAX_SUMMARY_MONTH_RANGE} months`,
  })
  dateTo: string;
}
