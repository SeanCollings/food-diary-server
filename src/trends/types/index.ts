import { IsIn, IsNotEmpty, IsString } from 'class-validator';
import { WellnessTypesEnum } from '@/wellness/types';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export type TrendType = 'week' | 'month';

export class GetTrendsQuery {
  @IsString()
  @IsNotEmpty()
  @IsIn(['week', 'month'])
  type: TrendType;
}

export type BeverageTypes = Exclude<
  WellnessTypesEnum,
  WellnessTypesEnum.EXCERCISE
>;
