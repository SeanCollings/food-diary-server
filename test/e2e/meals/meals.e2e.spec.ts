import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';
import { PrismaClient } from '@prisma/client';
import { PrismaService } from '@/prisma.service';
import { AppModule } from '@/app.module';
import { AuthModule } from '@/auth/auth.module';
import { UsersService } from '@/users/users.service';
import { isValidPassword } from '@/lib/validation/validate-user';
import { GoogleAdapter } from '@/auth/adapter/google.adapter';
import { UsersModule } from '@/users/users.module';
import { MealTypesEnum } from '@/meals/types';

import { mockLoginResponseBody } from '../e2e.setup';
import {
  diaryDaySingle_fixture,
  mealContentDto_fixture,
  user_fixture,
} from '../__fixtures__';
import { todayDate_fixture } from '../__fixtures__/today.date.fixture';

jest.mock('@/lib/validation/validate-user');

const mockError = jest.fn();

describe('SummaryController (e2e)', () => {
  let app: INestApplication;
  let prisma: DeepMockProxy<PrismaClient>;
  let mockUsersService: jest.Mocked<Partial<UsersService>>;
  let mockGoogleAdapter: jest.Mocked<Partial<GoogleAdapter>>;

  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(mockError);
  });

  beforeEach(async () => {
    mockUsersService = {
      findOne: jest.fn(),
    };
    mockGoogleAdapter = {
      verifySite: jest.fn(),
    };

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule, AuthModule, UsersModule],
    })
      .overrideProvider(PrismaService)
      .useValue(mockDeep<PrismaClient>())
      .overrideProvider(GoogleAdapter)
      .useValue(mockGoogleAdapter)
      .compile();

    app = moduleFixture.createNestApplication();
    await app.init();
    prisma = moduleFixture.get(PrismaService);
    mockGoogleAdapter = moduleFixture.get(GoogleAdapter);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('/ (POST)', () => {
    let access_token: string;

    beforeEach(async () => {
      (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
      (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
      (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

      access_token = (await mockLoginResponseBody(request, app)).access_token;

      prisma.user.findUnique.mockResolvedValue(user_fixture);
      prisma.user.update.mockResolvedValue(user_fixture);
    });

    describe('positive', () => {
      it('should create a new meal entry', async () => {
        prisma.diaryDay.findUnique.mockResolvedValue(diaryDaySingle_fixture);
        prisma.diaryDay.upsert.mockResolvedValue({} as any);

        const response = await request(app.getHttpServer())
          .post('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .query({ date: todayDate_fixture })
          .send({
            mealId: MealTypesEnum.BREAKFAST,
            content: mealContentDto_fixture,
          });

        expect(response.status).toEqual(201);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('negative', () => {
      it('should not allow non-logged in users', async () => {
        const response = await request(app.getHttpServer()).post('/meals');

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });

      it('should cater for lack of date query', async () => {
        const response = await request(app.getHttpServer())
          .post('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .send({
            mealId: MealTypesEnum.BREAKFAST,
            content: mealContentDto_fixture,
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });

      it('should cater for lack of body', async () => {
        const response = await request(app.getHttpServer())
          .post('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .query({ date: todayDate_fixture });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });

      it('should cater for bad content format', async () => {
        const response = await request(app.getHttpServer())
          .post('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .query({ date: todayDate_fixture })
          .send({
            mealId: MealTypesEnum.BREAKFAST,
            content: {},
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe('/ (PUT)', () => {
    let access_token: string;

    beforeEach(async () => {
      (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
      (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
      (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

      access_token = (await mockLoginResponseBody(request, app)).access_token;

      prisma.user.findUnique.mockResolvedValue(user_fixture);
      prisma.user.update.mockResolvedValue(user_fixture);
    });

    describe('positive', () => {
      it('should update a meal entry', async () => {
        prisma.diaryDay.findUnique.mockResolvedValue(diaryDaySingle_fixture);
        prisma.diaryDay.update.mockResolvedValue({} as any);

        const response = await request(app.getHttpServer())
          .put('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .query({ date: todayDate_fixture })
          .send({
            oldMealId: MealTypesEnum.BREAKFAST,
            newMealId: MealTypesEnum.LUNCH,
            content: mealContentDto_fixture,
          });

        expect(response.status).toEqual(200);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('negative', () => {
      it('should not allow non-logged in users', async () => {
        const response = await request(app.getHttpServer()).put('/meals');

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });

      it('should cater for lack of date query', async () => {
        const response = await request(app.getHttpServer())
          .put('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .send({
            oldMealId: MealTypesEnum.BREAKFAST,
            newMealId: MealTypesEnum.LUNCH,
            content: mealContentDto_fixture,
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });

      it('should cater for lack of body', async () => {
        const response = await request(app.getHttpServer())
          .put('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .query({ date: todayDate_fixture });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });

      it('should cater for bad content format', async () => {
        const response = await request(app.getHttpServer())
          .put('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .query({ date: todayDate_fixture })
          .send({
            oldMealId: MealTypesEnum.BREAKFAST,
            newMealId: MealTypesEnum.LUNCH,
            content: {},
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe('/ (DELETE)', () => {
    let access_token: string;

    beforeEach(async () => {
      (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
      (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
      (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

      access_token = (await mockLoginResponseBody(request, app)).access_token;

      prisma.user.findUnique.mockResolvedValue(user_fixture);
      prisma.user.update.mockResolvedValue(user_fixture);
    });

    describe('positive', () => {
      it('should delete a meal entry', async () => {
        prisma.diaryDay.findUnique.mockResolvedValue(diaryDaySingle_fixture);
        prisma.diaryDay.update.mockResolvedValue({} as any);

        const response = await request(app.getHttpServer())
          .delete('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .query({ date: todayDate_fixture })
          .send({
            mealId: MealTypesEnum.BREAKFAST,
            id: '1234567890123',
          });

        expect(response.status).toEqual(200);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('negative', () => {
      it('should not allow non-logged in users', async () => {
        const response = await request(app.getHttpServer()).delete('/meals');

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });

      it('should cater for lack of date query', async () => {
        const response = await request(app.getHttpServer())
          .delete('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .send({
            mealId: MealTypesEnum.BREAKFAST,
            id: '1234567890123',
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });

      it('should cater for lack of body', async () => {
        const response = await request(app.getHttpServer())
          .delete('/meals')
          .set('Authorization', 'Bearer ' + access_token)
          .query({ date: todayDate_fixture });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });
    });
  });
});
