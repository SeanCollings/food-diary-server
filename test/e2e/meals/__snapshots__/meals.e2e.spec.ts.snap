// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealsController (e2e) / (DELETE) negative should cater for lack of body 1`] = `
{
  "error": "Bad Request",
  "message": [
    "mealId must be one of the following values: breakfast, snack_1, lunch, snack_2, dinner",
    "mealId should not be empty",
    "mealId must be a string",
    "id must be shorter than or equal to 13 characters",
    "id must be longer than or equal to 13 characters",
    "id should not be empty",
    "id must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`MealsController (e2e) / (DELETE) negative should cater for lack of date query 1`] = `
{
  "error": "Bad Request",
  "message": [
    "date must be a valid ISO 8601 date string",
    "date should not be empty",
  ],
  "statusCode": 400,
}
`;

exports[`MealsController (e2e) / (DELETE) negative should not allow non-logged in users 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`MealsController (e2e) / (DELETE) positive should delete a meal entry 1`] = `{}`;

exports[`MealsController (e2e) / (POST) negative should cater for bad content format 1`] = `
{
  "error": "Bad Request",
  "message": [
    "content.id must be shorter than or equal to 13 characters",
    "content.id must be longer than or equal to 13 characters",
    "content.id should not be empty",
    "content.id must be a string",
    "content.food must be shorter than or equal to 100 characters",
    "content.food should not be empty",
    "content.food must be a string",
    "content.serving must be shorter than or equal to 100 characters",
    "content.serving must be a string",
    "content.quantity must be shorter than or equal to 100 characters",
    "content.quantity must be a string",
    "content.description must be shorter than or equal to 256 characters",
    "content.description must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`MealsController (e2e) / (POST) negative should cater for lack of body 1`] = `
{
  "error": "Bad Request",
  "message": [
    "mealId must be one of the following values: breakfast, snack_1, lunch, snack_2, dinner",
    "mealId should not be empty",
    "mealId must be a string",
    "content should not be empty",
  ],
  "statusCode": 400,
}
`;

exports[`MealsController (e2e) / (POST) negative should cater for lack of date query 1`] = `
{
  "error": "Bad Request",
  "message": [
    "date must be a valid ISO 8601 date string",
    "date should not be empty",
  ],
  "statusCode": 400,
}
`;

exports[`MealsController (e2e) / (POST) negative should not allow non-logged in users 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`MealsController (e2e) / (POST) positive should create a new meal entry 1`] = `{}`;

exports[`MealsController (e2e) / (PUT) negative should cater for bad content format 1`] = `
{
  "error": "Bad Request",
  "message": [
    "content.id must be shorter than or equal to 13 characters",
    "content.id must be longer than or equal to 13 characters",
    "content.id should not be empty",
    "content.id must be a string",
    "content.food must be shorter than or equal to 100 characters",
    "content.food should not be empty",
    "content.food must be a string",
    "content.serving must be shorter than or equal to 100 characters",
    "content.serving must be a string",
    "content.quantity must be shorter than or equal to 100 characters",
    "content.quantity must be a string",
    "content.description must be shorter than or equal to 256 characters",
    "content.description must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`MealsController (e2e) / (PUT) negative should cater for lack of body 1`] = `
{
  "error": "Bad Request",
  "message": [
    "newMealId must be one of the following values: breakfast, snack_1, lunch, snack_2, dinner",
    "newMealId should not be empty",
    "newMealId must be a string",
    "oldMealId must be one of the following values: breakfast, snack_1, lunch, snack_2, dinner",
    "oldMealId should not be empty",
    "oldMealId must be a string",
    "content should not be empty",
  ],
  "statusCode": 400,
}
`;

exports[`MealsController (e2e) / (PUT) negative should cater for lack of date query 1`] = `
{
  "error": "Bad Request",
  "message": [
    "date must be a valid ISO 8601 date string",
    "date should not be empty",
  ],
  "statusCode": 400,
}
`;

exports[`MealsController (e2e) / (PUT) negative should not allow non-logged in users 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`MealsController (e2e) / (PUT) positive should update a meal entry 1`] = `{}`;
